Add investment cost

lindnemi opened the pull request:
For Ariadne, but probably also useful for other people.

Adding investment directly to the model. Should be easier than getting it in the post-processing

lindnemi commented:
- [ ] Investments for grid connection should be added to `investment`
- [ ] Investments for transmission should be reported as a separate entry `transmission_investment` (for offwind)

lindnemi commented:
Seems to be working but needs a thorough review. Together with the corresponding changes in pypsa ariadne and pypsa-pypsa it's easy to get nice statistics on investments, e.g.

![image](https://github.com/user-attachments/assets/a1a14718-da25-4ad0-9203-76154dfff29a)

lindnemi commented:
- [ ] For consistency i should als add `connection_investment` for offwind in add_electricity. Will do that tomorrow.

fneum commented on doc/release_notes.rst:
```diff
@@ -11,6 +11,8 @@ Release Notes
 Upcoming Release
 ================
 
+* Added `investment` column to all components. These are the total upfront investments. The column is intended only for reporting and has no effect on the optimisation. For offshore wind, the `investment` column contains only investments into the wind park, the `connection_investment` is reported in a separate column.
```
Can you add this also to the `doc/configtables/costs.csv` that the investment column has no effect on optimisation?

fneum commented on config/config.default.yaml:
```diff
@@ -769,6 +769,13 @@ clustering:
       committable: any
       ramp_limit_up: max
       ramp_limit_down: max
+      investment: mean
+      connection_investment: mean
+    lines:
+      investment: mean
+    one_ports:
+      investment: mean
```
one_ports? or explicitly list generators

fneum commented on scripts/add_electricity.py:
```diff
@@ -352,6 +360,9 @@ def update_transmission_costs(n, costs, length_factor=1.0):
     n.lines["capital_cost"] = (
         n.lines["length"] * length_factor * costs.at["HVAC overhead", "capital_cost"]
     )
+    n.lines["investment"] = (
```
I am contemplating whether in PyPSA network the column should be called `overnight_cost`. This term might be less ambiguous than `investment`. For cleaning the `technology-data` CSV files, it is clear that it should be named `investment`.

fneum commented on scripts/add_electricity.py:
```diff
@@ -406,18 +430,32 @@ def attach_wind_and_solar(
             else:
                 capital_cost = costs.at[car, "capital_cost"]
+                investment = costs.at[car, "investment"]
+                connection_investment = None
```
Perhaps 0 would be a sensible default value or `pd.NA/np.nan`. `None` is not a good choice I think.

fneum commented on scripts/add_electricity.py:
```diff
@@ -507,6 +547,7 @@ def attach_conventional_generators(
         efficiency=ppl.efficiency,
         marginal_cost=marginal_cost,
         capital_cost=ppl.capital_cost,
+        # Not clear how to get investment for these Generators, hence ignoring it for now
```
```suggestion
        investment=ppl.investment,
```
I think it's already there through previous lines

```py
ppl = (
        ppl.query("carrier in @carriers")
        .join(costs, on="carrier", rsuffix="_r")
        .rename(index=lambda s: f"C{str(s)}")
)
```

fneum commented on scripts/add_electricity.py:
```diff
@@ -793,9 +837,16 @@ def attach_line_rating(
 
 if __name__ == "__main__":
     if "snakemake" not in globals():
+        import sys
+
         from _helpers import mock_snakemake
 
-        snakemake = mock_snakemake("add_electricity")
+        path = "../submodules/pypsa-eur/scripts"
+        sys.path.insert(0, os.path.abspath(path))
+        snakemake = mock_snakemake(
+            "add_electricity",
+            run="KN2045_Bal_v4",
+        )
```
revert

fneum commented on scripts/prepare_sector_network.py:
```diff
@@ -683,6 +717,9 @@ def add_co2_tracking(n, costs, options):
         e_nom_extendable=True,
         e_nom_max=e_nom_max,
         capital_cost=options["co2_sequestration_cost"],
+        investment=options[
+            "co2_sequestration_cost"
+        ],  # Just filling in capital_cost here
```
Should be either a correct value or NaN

fneum commented on scripts/prepare_sector_network.py:
```diff
@@ -750,6 +800,8 @@ def add_allam(n, costs):
         p_nom_extendable=True,
         # TODO: add costs to technology-data
         capital_cost=0.6 * 1.5e6 * 0.1,  # efficiency * EUR/MW * annuity
+        # TODO: add investment
+        # investment=0.6 * 1.5e6,  # efficiency * EUR/MW
```
Correct.

lindnemi replied:
credits to CoPilot for writing that comment :D

fneum commented on scripts/prepare_sector_network.py:
```diff
@@ -1232,12 +1313,18 @@ def add_storage_and_grids(n, costs):
             gas_pipes["p_nom_min"] = 0.0
             # 0.1 EUR/MWkm/a to prefer decommissioning to address degeneracy
             gas_pipes["capital_cost"] = 0.1 * gas_pipes.length
+            gas_pipes["investment"] = (
+                np.nan
+            )  # I guess with this option no investments occur??
```
Since it's only descriptive, I would say NaN.

fneum commented on scripts/prepare_sector_network.py:
```diff
@@ -2228,6 +2350,7 @@ def add_heat(n, costs):
                     country=ct,
                     capital_cost=capital_cost[strength]
                     * options["retrofitting"]["cost_factor"],
+                    # TODO add investment
```
Open TODO (commenting for tracking in PR)

fneum reviewed (changes requested):
Overall, I think this works relatively well, even though it is quite a bit of duplication.

My main thought was wether `overnight_cost` might be more suitable for the attribute name in PyPSA networks.

lindnemi commented:
Addressed most of your comments, thanks for the review.

Retrofitting cost not yet added, since we do not use it atm and i wouldn't know how to find a suitable value.

I agree that `overnight_cost` is more concise than `investment`, however since the terminology is the same in technology data i am fine with leaving it like it is. 

If you prefer `overnight_cost`, just let me know and i will change it

fneum commented:
> If you prefer overnight_cost, just let me know and i will change it

Yes, I have a preference for `overnight_cost`. Some divergence from `technology-data? is fine (`prepare_sector_network` also uses "fixed" for "capital_cost")

fneum commented:
TODO cherry-pick squashed commit into PyPSA-Eur `master`
